### PR TITLE
Start optimising the memory footprint of the UTxO.

### PR DESCRIPTION
**Pull request by kantp**
This changes the memory representation of TxIn and TxOut, to reduce overhead.

In particular, we use unpacked Word64 for coin values and txout indices, and use
the serialisation of addresses as ShortByteString.

This does not yet change the representation of TxId.

The changes are nearly opaque, in that we provide view patterns that takes the
place of the old constructors. We do need to add a Crypto constraint wherever we
use a TxOut, however, since the view pattern relies on
serialisation/deserialisation.

This is a draft PR for now, as I'd like to add some properties around the TxOut (I've introduced a `panic` when deserialisation fails in the view pattern, which should never occur, since the constructor itself is not exported, and so the only way to construct a TxOut is using serialisation).

### CODE REVIEW

**Comment by dcoutts** [comment]
Since we rely on the (de)serialiseAddress then we should inherit the round trip property, but it's worth adding a QC test anyway.

**Review by JaredCorduan — APPROVED** [review]
LGTM!

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs [hunk 1]
```diff
@@ -375,18 +384,59 @@ deriving newtype instance Crypto crypto => ToCBOR (TxId crypto)
 deriving newtype instance Crypto crypto => FromCBOR (TxId crypto)
 
 -- | The input of a UTxO.
-data TxIn crypto
-  = TxIn !(TxId crypto) !Natural -- TODO use our own Natural type
+data TxIn crypto = TxInCompact !(TxId crypto) {-# UNPACK #-} !Word64
```
Comment (dcoutts) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs]:
```suggestion
data TxIn crypto = TxInCompact {-# UNPACK #-} !(TxId crypto) {-# UNPACK #-} !Word64
```
The `TxId` is a single constructor type so we can unpack that one too.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs [hunk 2]
```diff
@@ -375,18 +384,59 @@ deriving newtype instance Crypto crypto => ToCBOR (TxId crypto)
...
   deriving (Show, Eq, Generic, Ord, NFData)
 
+-- TODO: We will also want to have the TxId be compact, but the representation
+-- depends on the crypto.
+
+pattern TxIn ::
+  Crypto crypto =>
+  TxId crypto ->
+  Natural -> -- TODO We might want to change this to Word64 generally
```
Comment (dcoutts) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:396]: I agree with this TODO.

**Review by dcoutts — APPROVED** [review]
LGTM!

**Pull request merged**